### PR TITLE
gitstatus: add prompt scripts and share helper 

### PR DESCRIPTION
**Pull request by chetgurevitch**
## Description of changes

Packages the prompt scripts to use gitstatus directly in bash and zsh. Also adds a gitstatus-share script to more easily source them and updated installation instructions in the longDescription. I have an [overhaul](https://github.com/chetgurevitch/nixpkgs/commit/d5a53f887a572a372c1aca407b843a6aac463e12) of the shell-helper docs to PR afterwards.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/version-management/gitstatus/default.nix [hunk 1]
```diff
@@ -83,9 +102,26 @@ stdenv.mkDerivation rec {
...
     description = "10x faster implementation of `git status` command";
+    longDescription = ''
+      To enable a prompt using gitstatus directly, add one of the following to your NixOS configuration:
+      `programs.bash.promptInit = "source $(gitstatus-share)/gitstatus.prompt.sh";`
+      `programs.zsh.promptInit = "source $(gitstatus-share)/gitstatus.prompt.zsh";`
+
+      To enable from Home Manager:
+      `programs.bash.initExtra = "source $(gitstatus-share)/gitstatus.prompt.sh";`
+      `programs.zsh.initExtra = "source $(gitstatus-share)/gitstatus.prompt.zsh";`
+
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitstatus/default.nix:113]:
```suggestion
```
We won't notice in nixpkgs when this changes in home-manager.

File: pkgs/applications/version-management/gitstatus/default.nix [hunk 2]
```diff
@@ -23,13 +30,23 @@ stdenv.mkDerivation rec {
...
     # because the FHS directories don't start at /
     substituteInPlace install \
       --replace "_gitstatus_install_main ." "_gitstatus_install_main $out"
+
+    cat <<SCRIPT > $out/bin/gitstatus-share
+    #!${runtimeShell}
+    # Run this script to find the gitstatus shared folder where all the shell
+    # integration scripts are living.
+    echo $out/share/gitstatus
+    SCRIPT
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitstatus/default.nix:48]:
```suggestion
    cat <<EOF > $out/bin/gitstatus-share
    #!${runtimeShell}
    # Run this script to find the gitstatus shared folder where all the shell
    # integration scripts are living.
    echo $out/share/gitstatus
    EOF
```

File: pkgs/applications/version-management/gitstatus/default.nix [hunk 3]
```diff
@@ -11,7 +18,7 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-b+9bwJ87VV6rbOPobkwMkDXGH34STjYPlt8wCRR5tEc=";
   };
 
-  buildInputs = [ (callPackage ./romkatv_libgit2.nix { }) ];
+  buildInputs = [(callPackage ./romkatv_libgit2.nix {})];
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/gitstatus/default.nix:21]: We are using nixfmt-rfc-style for formatting which undo this again.

```
diff --git a/pkgs/applications/version-management/gitstatus/default.nix b/pkgs/applications/version-management/gitstatus/default.nix
index 931acbefb597..1d885bd5bb42 100644
--- a/pkgs/applications/version-management/gitstatus/default.nix
+++ b/pkgs/applications/version-management/gitstatus/default.nix
@@ -18,7 +18,7 @@
     sha256 = "sha256-b+9bwJ87VV6rbOPobkwMkDXGH34STjYPlt8wCRR5tEc=";
   };

-  buildInputs = [(callPackage ./romkatv_libgit2.nix {})];
+  buildInputs = [ (callPackage ./romkatv_libgit2.nix { }) ];

   postPatch = ''
     sed -i '1i GITSTATUS_AUTO_INSTALL=''${GITSTATUS_AUTO_INSTALL-0}' gitstatus.plugin.sh
@@ -64,7 +64,10 @@
   # that the script was sourced successfully and that the "gitstatus_query"
   # command ran successfully. This tests the binary itself and the zsh
   # integration.
-  nativeInstallCheckInputs = [git zsh];
+  nativeInstallCheckInputs = [
+    git
+    zsh
+  ];
   doInstallCheck = true;
   installCheckPhase = ''
     TEMP=$(mktemp -d)
@@ -121,7 +124,11 @@
     '';
     homepage = "https://github.com/romkatv/gitstatus";
     license = licenses.gpl3Only;
-    maintainers = with maintainers; [mmlb hexa SuperSandro2000];
+    maintainers = with maintainers; [
+      mmlb
+      hexa
+      SuperSandro2000
+    ];
     platforms = platforms.all;
     mainProgram = "gitstatusd";
   };
```

File: pkgs/applications/version-management/gitstatus/default.nix (hunk 1 above)
Comment (chetgurevitch) [reply pkgs/applications/version-management/gitstatus/default.nix]: Removed

File: pkgs/applications/version-management/gitstatus/default.nix (hunk 2 above)
Comment (chetgurevitch) [reply pkgs/applications/version-management/gitstatus/default.nix]: Changed SCRIPT to EOF

File: pkgs/applications/version-management/gitstatus/default.nix [hunk 4]
```diff
@@ -83,9 +102,26 @@ stdenv.mkDerivation rec {
...
+      `programs.bash.initExtra = "source $(gitstatus-share)/gitstatus.prompt.sh";`
+      `programs.zsh.initExtra = "source $(gitstatus-share)/gitstatus.prompt.zsh";`
+
+      Afterwards you can customize your prompt using the `$PS1` environment variable for Bash or the `$PROMPT` and
+      `$RPROMPT` variables for Zsh, simply include `$GITSTATUS_PROMPT` where desired. For control over the format of
+      gitstatus you can instead source the `gitstatus.plugin.{sh,zsh}` shell script directly then copy from and modify
+      the prompt manually, see the project homepage for more details.
+
+      If you'd like to preserve your prompt when dropping into a nix-shell you can set the `NIX_SHELL_PRESERVE_PROMPT`
+      environment variable to 1.
```
[the remark this answers is not in the document]
Comment (chetgurevitch) [reply pkgs/applications/version-management/gitstatus/default.nix]: Removed

File: pkgs/applications/version-management/gitstatus/default.nix (hunk 3 above)
Comment (chetgurevitch) [reply pkgs/applications/version-management/gitstatus/default.nix]: Reformatted with nixfmt

File: pkgs/applications/version-management/gitstatus/default.nix [hunk 5]
```diff
@@ -83,9 +102,26 @@ stdenv.mkDerivation rec {
...
+      To enable a prompt using gitstatus directly, add one of the following to your NixOS configuration:
+      `programs.bash.promptInit = "source $(gitstatus-share)/gitstatus.prompt.sh";`
+      `programs.zsh.promptInit = "source $(gitstatus-share)/gitstatus.prompt.zsh";`
+
+      To enable from Home Manager:
+      `programs.bash.initExtra = "source $(gitstatus-share)/gitstatus.prompt.sh";`
+      `programs.zsh.initExtra = "source $(gitstatus-share)/gitstatus.prompt.zsh";`
+
+      Afterwards you can customize your prompt using the `$PS1` environment variable for Bash or the `$PROMPT` and
+      `$RPROMPT` variables for Zsh, simply include `$GITSTATUS_PROMPT` where desired. For control over the format of
```
[the remark this answers is not in the document]
Comment (chetgurevitch) [reply pkgs/applications/version-management/gitstatus/default.nix]: Replaced with a recommendation to see the project homepage for details on customization

**Comment by chetgurevitch** [comment]
@SuperSandro2000 any remaining concerns?

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**